### DISCUSSION TITLE
How to serialize json in ELSA

### DISCUSSION THREAD

**Discussion by RekikMohamed** [elsa-workflows/elsa-core#6045] [Q&A]
I tried the sample code in the ELSA documentation to serialize a workflow from json.
```
using Elsa.Extensions;
using Elsa.Testing.Shared;
using Elsa.Workflows.Core.Contracts;
using Elsa.Workflows.Core.Models;
using Microsoft.Extensions.DependencyInjection;

// Setup service container.
var services = new ServiceCollection();

// Add Elsa services.
services.AddElsa();

// Build service container.
var serviceProvider = services.BuildServiceProvider();

// Populate registries. This is only necessary for applications  that are not using hosted services.
await serviceProvider.PopulateRegistriesAsync();

// Import a workflow from a JSON file.
var workflowJson = await File.ReadAllTextAsync("HelloWorld.json");

// Get a serializer to deserialize the workflow.
var serializer = serviceProvider.GetRequiredService<IActivitySerializer>();

// Deserialize the workflow.
var workflow = serializer.Deserialize<Workflow>(workflowJson);

// Resolve a workflow runner to run the workflow.
var workflowRunner = serviceProvider.GetRequiredService<IWorkflowRunner>();

// Run the workflow.
await workflowRunner.RunAsync(workflow);
```

The PopulateRegistriesAsync() method and some references do not exist, I get the following errors:

- Error	CS0234	The type or namespace name 'Testing' does not exist in the namespace 'Elsa' (are you missing an assembly reference?)	
- Error	CS0234	The type or namespace name 'Core' does not exist in the namespace 'Elsa.Workflows' (are you missing an assembly reference?)	
- Error	CS0234	The type or namespace name 'Core' does not exist in the namespace 'Elsa.Workflows' (are you missing an assembly reference?)
- Error	CS1061	'ServiceProvider' does not contain a definition for 'PopulateRegistriesAsync' and no accessible extension method 'PopulateRegistriesAsync' accepting a first argument of type 'ServiceProvider' could be found (are you missing a using directive or an assembly reference?)

When I remove the unfound references and run the application, I get the following exception stack in the deserialization:

   at Elsa.Workflows.Services.ActivityFactory.ReadSyntheticInputs(ActivityDescriptor activityDescriptor, IActivity activity, JsonElement activityRoot, JsonSerializerOptions options)
   at Elsa.Workflows.Services.ActivityFactory.Create(Type type, ActivityConstructorContext context)
   at Elsa.Workflows.Serialization.Converters.ActivityJsonConverter.Read(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.ReadJsonAndSetMember(Object obj, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 utf8Json, JsonTypeInfo`1 jsonTypeInfo, Nullable`1 actualByteCount)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 json, JsonTypeInfo`1 jsonTypeInfo)
   at System.Text.Json.JsonSerializer.Deserialize[TValue](String json, JsonSerializerOptions options)
   at Program.<<Main>$>d__0.MoveNext() in D:\repos\WorkflowEngine\ElsaWorkflow\Program.cs:line 24
   
  
Any idea why is this happening?

**Comment by zhenlee-lt** [comment]
Any updates? I had the same issue. It's not working following the documentation.

**Comment by Green7** [comment]
You need to correct the "uses" section:

```using Elsa.Extensions;
using Elsa.Testing.Shared;
using Elsa.Workflows;
using Elsa.Workflows.Activities;
using Microsoft.Extensions.DependencyInjection;
```

**Comment by hauke-w** [comment]
> * Error	CS1061	'ServiceProvider' does not contain a definition for 'PopulateRegistriesAsync' and no accessible extension method 'PopulateRegistriesAsync' accepting a first argument of type 'ServiceProvider' could be found (are you missing a using directive or an assembly reference?)

I solved this by adding the Nuget package `Elsa.Testing.Shared.Integration`.

**Comment by sfmskywalker** [comment]
What is happening here is a mix of outdated namespaces/docs and one helper that lives in the integration-testing package.

From the current codebase:

- `Workflow` is in `Elsa.Workflows.Activities`, not `Elsa.Workflows.Core.Models`:
  `/Users/sipke/Projects/Elsa/elsa-core/src/modules/Elsa.Workflows.Core/Activities/Workflow.cs`
- `IWorkflowRunner` and the core contracts are in the `Elsa.Workflows` namespace:
  `/Users/sipke/Projects/Elsa/elsa-core/src/modules/Elsa.Workflows.Core/Services/WorkflowRunner.cs`
- `PopulateRegistriesAsync()` is an extension method provided by the integration test helper package:
  `/Users/sipke/Projects/Elsa/elsa-core/src/common/Elsa.Testing.Shared.Integration/ServiceProviderExtensions.cs`
- under the hood, that helper resolves `IRegistriesPopulator` and calls `PopulateAsync()`:
  `/Users/sipke/Projects/Elsa/elsa-core/src/modules/Elsa.Workflows.Runtime/Contracts/IRegistriesPopulator.cs`

So the current imports should be closer to:

```csharp
using Elsa.Extensions;
using Elsa.Workflows;
using Elsa.Workflows.Activities;
using Microsoft.Extensions.DependencyInjection;
```

And if you want to use `PopulateRegistriesAsync()` exactly as shown, you need the `Elsa.Testing.Shared.Integration` package/reference, because that extension method is not part of the main runtime surface.

If you do **not** want to pull in the testing package, the runtime equivalent is to resolve `IRegistriesPopulator` and call `PopulateAsync()` yourself before deserializing JSON-defined workflows.

So the key points are:

- `Elsa.Workflows.Core.*` namespaces from the old sample are no longer the right imports
- `PopulateRegistriesAsync()` is a testing helper, not a core runtime extension
- deserialization failures often happen when the activity registries were not populated before reading the JSON

Relevant source:
- `elsa-core/src/modules/Elsa.Workflows.Core/Activities/Workflow.cs`
- `elsa-core/src/modules/Elsa.Workflows.Core/Services/WorkflowRunner.cs`
- `elsa-core/src/common/Elsa.Testing.Shared.Integration/ServiceProviderExtensions.cs`
- `elsa-core/src/modules/Elsa.Workflows.Runtime/Contracts/IRegistriesPopulator.cs`
- `elsa-core/src/modules/Elsa.Workflows.Runtime/Services/DefaultRegistriesPopulator.cs`

So yes: the documentation snippet is stale in its current form.
